File: blackjack/ruby/outcomes.rb
#!/usr/bin/env ruby

def partitions(cards, subtotal)
  m = 0
  # Hit
  for i in 0..9
    if (cards[i]>0) then
      total = subtotal+i+1
      if (total < 21) then
        # Stand
        m += 1
        # Hit again
        cards[i] += -1
        m += partitions(cards, total)
        cards[i] += 1
      elsif (total==21) then
        # Stand; hit again is an automatic bust
        m += 1
        break
      end
    end
  end
  return m
end

deck = ([4]*9)
deck << 16
d = 0

for i in 0..9
  # Dealer showing
  deck[i] -= 1
  p = 0
  for j in 0..9
    deck[j] -= 1
    n = partitions(deck, j+1)
    deck[j] += 1
    p += n
  end
  print("Dealer showing #{i} partitions = #{p}\n")
  d += p
  deck[i] += 1
end

print("Total partitions = #{d}\n")

File: blackjack/perl/outcomes.pl
sub partitions(@$)
{
    my @cards = @{$_[0]};
    my $subtotal = $_[1];
    my $m = 0;
    # Hit
    for(my $i = 0; $i<10; $i++) {
	if ($cards[$i]>0) {
            my $total = $subtotal+$i+1;
            if ($total < 21) {
                # Stand
                $m += 1;
                # Hit again
		$cards[$i] = $cards[$i]-1;
                $m += partitions(\@cards,$total);
                $cards[$i] = $cards[$i]+1;
            } elsif ($total==21) {
                # Stand; hit again is an automatic bust
                $m += 1;
		break;
            }
	}
    }
    return $m;
}

my $d = 0;
my @deck = (4,4,4,4,4,4,4,4,4,16);

for(my $i = 0; $i<10; $i++) {
    # Dealer showing
    $deck[$i] = $deck[$i]-1;
    my $p = 0;
    for(my $j = 0; $j<10; $j++) {
        $deck[$j] = $deck[$j]-1;
        my $n = partitions(\@deck,$j+1);
        $deck[$j] = $deck[$j]+1;
        $p += $n;
    }

    print "Dealer showing ",$i," partitions = ",$p,"\n";
    $d += $p;

    $deck[$i] = $deck[$i]+1;
}

print "Total partitions = ",$d,"\n";
